services/validators/envValidator.go

```diff
@@ -151,6 +151,60 @@ func closeDBConnection(handle *sql.DB) {
 	}
 }
 
+func killDanglingDBConnections(db *sql.DB) {
+
+	rows, err := db.Query(`SELECT PID, QUERY_START, WAIT_EVENT_TYPE, WAIT_EVENT, STATE, QUERY
+							FROM PG_STAT_ACTIVITY
+							WHERE PID <> PG_BACKEND_PID()
+							AND APPLICATION_NAME = (
+								SELECT APPLICATION_NAME FROM PG_STAT_ACTIVITY WHERE PID = PG_BACKEND_PID()
+							)`)
+
+	if err != nil {
+		panic(fmt.Errorf("error occurred with querying pg_stat_activity table for finding dangling connections: %v", err.Error()))
+	}
+	defer rows.Close()
+
+	type danglingConnRowT struct {
+		pid           int
+		queryStart    string
+		waitEventType string
+		waitEvent     string
+		state         string
+		query         string
+	}
+	type danglingConnsT []*danglingConnRowT
+
+	dangling := danglingConnsT(make([]*danglingConnRowT, 0))
+	for rows.Next() {
+		var row danglingConnRowT = danglingConnRowT{}
+		err := rows.Scan(&row.pid, &row.queryStart, &row.waitEventType, &row.waitEvent, &row.state, &row.query)
+		if err != nil {
+			panic(err)
+		}
+		dangling = append(dangling, &row)
+	}
+
+	if len(dangling) > 0 {
+		pkgLogger.Warn(fmt.Sprintf("Killing %d dangling connection(s)", len(dangling)))
+		for i, rowPtr := range dangling {
+			pkgLogger.Warn(fmt.Sprintf("dangling connection #%d: %+v", i+1, *rowPtr))
+		}
+
+		_, err := db.Exec(`SELECT PG_TERMINATE_BACKEND(PID)
```

Comment:
I don't understand/see the benefit of making two queries:
1. `db.Query(`SELECT PID ...`
2. `db.Exec(`SELECT PG_TERMINATE_BACKEND`

How the data from the first query is used?

Reply:
We should definitely log them. But running it twice can mean we might be logging some connections and killing different set of connections as connections might potentially change from the first query to the second. We can just iterate over the fetched PIDs and kill them explicitly. 

OR

Single query to kill and fetch the data might be best as suggested by Leo

Reply:
Also, did we consider PG_CANCEL_BACKEND instead of PG_TERMINATE_BACKEND?

Reply:
We should be doing it in a single query as @snarkychef and @lvrach mentioned.

FYI , We are using the below query on production to kill the dangling connections on postgres
```
SELECT
	pg_terminate_backend(pg_stat_activity.pid)
FROM
	pg_stat_activity
WHERE
	pg_stat_activity.datname = 'jobsdb'
	AND pid <> pg_backend_pid();
```

Reply:
> Also, did we consider PG_CANCEL_BACKEND instead of PG_TERMINATE_BACKEND?

`PG_CANCEL_BACKEND` cancels the query whereas `PG_TERMINATE_BACKEND` terminates the whole database connection. In our case, since this connection is dangling and nobody actually holds a handle on it, `PG_TERMINATE_BACKEND` appears to be more appropriate.

Reply:
> I don't understand/see the benefit of making two queries:
> 
> 1. `db.Query(`SELECT PID ...`
> 2. `db.Exec(`SELECT PG_TERMINATE_BACKEND`
> 
> How the data from the first query is used? Do we want to log them? If that is the case, could we have done that with a single query ?

Correct, will do both in the same query.

FYI, `PG_TERMINATE_BACKEND` also returns a `boolean` value indicating whether the termination was successful or not, however from what I understood the only occasion where this boolean value can be `false` is when you provide a PID which doesn't correspond to an actual database session. This is not expected to ever happen, since we will be terminating based on the `PG_STAT_ACTIVITY` view.